### ISSUE TITLE
Missing Python packages due to cleared `$PYTHONPATH`

### ISSUE THREAD

**Issue by Flamefire** [spyder-ide/spyder#22681]
## Problem Description

We install Spyder alongside other globally installed Python packages that are in non-standard locations. To make Python find them we either add the locations to `$PYTHONPATH` or use a custom `sitecustomize.py` whose location must be in `$PYTHONPATH`

When starting Spyder it fails with errors such as `No module named 'zmq'`.

I traced that to https://github.com/spyder-ide/spyder/blob/a857397bb4fc80c597ae3ba1ac478765137a1117/spyder/app/start.py#L13-L18

That removes the entries that are required such that Python finds the modules and our `sitecustomize`

Removing that results in a similar problem as the IDE console shows 
> Error while finding module specification for 'spyder_kernels.console' (ModuleNotFoundError: No module named 'spyder_kernels')

That arises due to removing `$PYTHONPATH` from the environment the kernel starts in:
https://github.com/spyder-ide/spyder/blob/a857397bb4fc80c597ae3ba1ac478765137a1117/spyder/plugins/ipythonconsole/utils/kernelspec.py#L172

I'm not fully sure what the intention was, but maybe that can be solved differently. Is that possible?

## Versions

* Spyder version: 5.5.1
* Python 3.11.3 64-bit | Qt 5.15.10 | PyQt5 5.15.10 | Linux 4.18.0-513.24.1.el8_9.x86_64 (x86_64)

**Comment by ccordoba12** [comment]
Hey @Flamefire, thanks for reporting. You said:

> We install Spyder alongside other globally installed Python packages that are in non-standard locations. To make Python find them we either add the locations to $PYTHONPATH or use a custom sitecustomize.py whose location must be in $PYTHONPATH

We haven't contemplated the possibility of people building a Python installation by adding different directories to `PYTHONPATH`.

> I'm not fully sure what the intention was

We remove `PYTHONPATH` before starting Spyder and from kernels to avoid users adding directories to it with modules named `random.py`, `string.py`, `math.py`, etc. Those modules would mask the respective standard library ones and cause major breakages. Most people are unaware that that is even possible and were very frustrated when it happened because it's not an easy problem to understand and find help about.

> but maybe that can be solved differently. Is that possible?

I don't think there is a simple way to prevent the problem I mentioned without dropping `PYTHONPATH` from the env. But I'm open to suggestions.

**Comment by Flamefire** [comment]
> We remove PYTHONPATH before starting Spyder and from kernels to avoid users adding directories to it

How exactly are they adding directories to it? IIRC I read in one of the related issues that it was about the current directory being used implicitly as a python search path. Maybe that could be detected and only that one removed from PYTHONPATH, potentially together with any empty values as IIRC `PYTHONPATH=/foo:` would mean the same as `/foo:$PWD`

**Comment by mrclary** [comment]
`PYTHONPATH` is for user-provided and non-standard-library paths and should not be relied upon for Python or _installed_ packages to work properly. Unfortunately, any user-provided path via `PYTHONPATH` presents a potential hazard to Spyder, not just the current working directory, so we cannot selectively "clean" this environment variable.

Rather than `PYTHONPATH`, I would recommend using a `.pth` file located in `../lib/python3.11/site-packages`, relative to the Python executable you are using.
See https://docs.python.org/3.11/library/site.html#module-site

**Comment by Flamefire** [comment]
Unfortunately that doesn't work for our use case: We have environment modules managed via [LMod](https://lmod.readthedocs.io/) to allow users to choose their desired version of software in a (heavily) shared environment/filesystem.

Those "modules" can be though of as (bash) scripts that are sourced and set/update and export environment variables. Hence the only thing a module providing a set of Python packages can do is append or prepend to `$PYTHONPATH`.

In particular we cannot modify the global Python installation and put additional files in there, so the `.pth` approach does not work for us.

We are switching to a mechanism using `sitecustomize.py` that reads a custom env variable to modify `sys.path` (as also described in your link). However we do put that into a separate folder and pass its location via `$PYTHONPATH`.


IIRC the `$PWD` is added to `sys.path` not `$PYTHONPATH`, I confused that. So any `$PYTHONPATH` would be something explicitly set by a user, so most "should" be aware of it.

As the intention is that people "were very frustrated when it happened because it's not an easy problem to understand and find help about." maybe another solution is to detect a set `$PYTHONPATH` and show a warning on startup to educate users that don't know about that but still allow this mechanism for users that do know and intended that (at least some like us)   
If there is a message like "You have set $PYTHONPATH. If you encounter errors related to imports of Python packages you can try unsetting that environment variable before starting Spyder" then user that are not aware of that have some help right there without taking away control from all users.

I can also imaging another variable like `$SPYDER_CLEAN_PYTHONPATH=0/false` to disable that which we can set in the module. That would be in line with the standard `$PYTHONNOUSERSITE` or `$PIP_REQUIRE_VIRTUALENV` variables.

**Comment by mrclary** [comment]
@Flamefire, thank you for the deeper explanation. I think I understand better now what your situation is. 

How much control do you have over these module scripts? 

Could you do something like the following? 
Assuming there is a module located at `/apps/modulefiles/mymodule/<ver>.lua`, create a `.pth` file in `/apps/modulefiles/mymodule/lib/python3.11/site-packages`, containing the necessary paths. Rather than setting the `PYTHONPATH` variable in the module, set the `PYTHONUSERBASE` environment variable to `/apps/modulefiles/mymodule`. After loading this module, running Python will result in `sys.path` containing all the paths listed in the `.pth` file, without needing to modify `PYTHONPATH`.

**Comment by Flamefire** [comment]
@mrclary Thanks for the reply and suggestions

> How much control do you have over these module scripts?

More or less full control. **BUT**:
I was reporting this in the context of EasyBuild that installs the software and creates the modules. So we can control the contents of the modules. However there are many existing modules already installed on multiple HPC clusters so changing those existing modules is virtually impossible.

> Rather than setting the `PYTHONPATH` variable in the module, set the `PYTHONUSERBASE` environment variable 

I didn't know about that variable and it might come in handy.   
However for our situation it doesn't help: There are multiple modules with Python packages. At the extreme it could be one module per package but in practice we have something like: 1 for Python (binaries etc.) and some base packages, 1 for commonly used packages from PyPI (~80 packages), 1 for numpy&scipy. And then some on top, e.g. h5py which requires HDF5 to be installed first, so it's sensible to put that in a extra modules in case some sites don't need it (and hence HDF5)

`PYTHONPATH` has the advantage of being a list. Hence we can prepend paths to this list in the same way we do with  `$PATH` and `$LD_LIBRARY_PATH`. 

So one solution I see is putting our `sitecustomize.py` into the `site-packages` folder and using that mechanism to read possible `sys.path` entries from our custom env variable, so we no longer require `$PYTHONPATH` for that.   
However that doesn't help with existing modules and we only recently introduced an option to use that as the default mechanism instead of `$PYTHONPATH`. But sites can still opt to use `$PYTHONPATH` for whatever reasons which would then cause that problem again for them.

And as confusing as it might be if users set up `$PYTHONPATH` and put modules/files inside those shadowing standard library ones it was very confusing for us to find the (by now 2) places where the PYTHONPATH we purposely set up was removed.  
Hence the idea of only showing a warning as a starting point for users that are not aware of that.

**Comment by mrclary** [comment]
> More or less full control. **BUT**: I was reporting this in the context of EasyBuild that installs the software and creates the modules. So we can control the contents of the modules. However there are many existing modules already installed on multiple HPC clusters so changing those existing modules is virtually impossible.

Yes, I see. That would require updating the EasyBuild and redeploy/reinstall across your clusters.

> > Rather than setting the `PYTHONPATH` variable in the module, set the `PYTHONUSERBASE` environment variable
> 
> I didn't know about that variable and it might come in handy. However for our situation it doesn't help: There are multiple modules with Python packages. At the extreme it could be one module per package but in practice we have something like: 1 for Python (binaries etc.) and some base packages, 1 for commonly used packages from PyPI (~80 packages), 1 for numpy&scipy. And then some on top, e.g. h5py which requires HDF5 to be installed first, so it's sensible to put that in a extra modules in case some sites don't need it (and hence HDF5)

Yes, I see. Python also provides `PYTHONHOME` for the purpose of using a non-standard location for the python libraries, but only for up to two locations and doesn't support scaling to 3 or more locations. In or order for `PYTHONUSERBASE` to work, each module would have to add/remove a `.pth` file to the `$PYTHONUSERBASE/lib/python3.11/site-packages` upon load/unload, which I don't think is possible.

> So one solution I see is putting our `sitecustomize.py` into the `site-packages` folder and using that mechanism to read possible `sys.path` entries from our custom env variable, so we no longer require `$PYTHONPATH` for that. 

I suppose that `sitecustomize.py` would require logic to determine which modules are loaded and modify `sys.path` accordingly.

> However that doesn't help with existing modules and we only recently introduced an option to use that as the default mechanism instead of `$PYTHONPATH`. But sites can still opt to use `$PYTHONPATH` for whatever reasons which would then cause that problem again for them.

Yep. Either the change is made with EasyBuild and redeploy/reinstall across your systems, or changes are made to Spyder and you update Spyder. How is Spyder installed in your situation? Does it have its own module or is it installed in the base Python location?

Python 3.11 introduced the `PYTHONSAFEPATH` which prevents prepending potentially unsafe paths to `sys.path`. I don't know to what extent this applies to paths in `PYTHONPATH` but nevertheless cannot be used since Spyder still supports Python <3.11; but it might be something to look at in the future.

So, in conclusion, I agree that @Flamefire's suggestion to use an environment variable to switch `PYTHONPATH` treatment in Spyder may be the best solution for now. I propose `SPY_ADD_PYTHONPATH`. Spyder will not add paths from `PYTHONPATH` to `sys.path` upon startup if `SPY_ADD_PYTHONPATH` is unset or evaluates to False; Spyder will add paths from `PYTHONPATH` to `sys.path` upon startup if `SPY_ADD_PYTHONPATH` evaluates to True. What do @spyder-ide/core-developers think?

**Comment by mrclary** [comment]
> In or order for `PYTHONUSERBASE` to work, each module would have to add/remove a `.pth` file to the `$PYTHONUSERBASE/lib/python3.11/site-packages` upon load/unload, which I don't think is possible.

Actually, I think this is possible, but still not convenient.

**Comment by Flamefire** [comment]
> Yes, I see. Python also provides `PYTHONHOME` for the purpose of using a non-standard location for the python libraries, but only for up to two locations and doesn't support scaling to 3 or more locations. In or order for `PYTHONUSERBASE` to work, each module would have to add/remove a `.pth` file to the `$PYTHONUSERBASE/lib/python3.11/site-packages` upon load/unload, which I don't think is possible.

Yes that is not possible because it is a shared filesystem, so that location needs to be at least specific to each user as they must not interfere with each other. But users can also have multiple independent tasks (from different shell sessions) running concurrently so that location must also be per-user-per-shell-session and still on the global filesystem which is impractical.

> I suppose that `sitecustomize.py` would require logic to determine which modules are loaded and modify `sys.path` accordingly.

Correct. We have a shell variable that is prepended to by modules and that file reads it and adds the appropriate paths to `sys.path`

> Yep. Either the change is made with EasyBuild and redeploy/reinstall across your systems, or changes are made to Spyder and you update Spyder. How is Spyder installed in your situation? Does it have its own module or is it installed in the base Python location?

Yes it has it's own module. So updating that is much easier, especially as currently it cannot be used as a module, although we have a workaround: We remove the unset-code with a patch. But a general solution avoiding the need for patches going forward would be appreciated.

> Python 3.11 introduced the `PYTHONSAFEPATH` which prevents prepending potentially unsafe paths to `sys.path`. I don't know to what extent this applies to paths in `PYTHONPATH`

That seems to avoid the common issue with "wrong" files next to the file you want to execute. Interesting option indeed.

> So, in conclusion, I agree that @Flamefire's suggestion to use an environment variable to switch `PYTHONPATH` treatment in Spyder may be the best solution for now. I propose `SPY_ADD_PYTHONPATH`. Spyder will not add paths from `PYTHONPATH` to `sys.path` upon startup if `SPY_ADD_PYTHONPATH` is unset or evaluates to False; Spyder will add paths from `PYTHONPATH` to `sys.path` upon startup if `SPY_ADD_PYTHONPATH` evaluates to True. What do @spyder-ide/core-developers think?

Just to be exact: Currently spider *unsets*  `$PYTHONPATH` and *removes* those from `sys.path` after the fact. So maybe an inverse name is better `SPY_CLEAN_PYTHONPATH` which defaults to True to keep the current behavior but can be set to False/0/Off e.g. in our module. Or default to False and show a warning with a hint that this can be used given that the cleaning is basically a workaround for user mistakes, not something strictly required.

**Comment by mrclary** [comment]
> Just to be exact: Currently spider unsets  $PYTHONPATH and removes those from sys.path after the fact. So maybe an inverse name is better SPY_CLEAN_PYTHONPATH which defaults to True to keep the current behavior but can be set to False/0/Off e.g. in our module. Or default to False and show a warning with a hint that this can be used given that the cleaning is basically a workaround for user mistakes, not something strictly required. So I'd use the more common case by default to have more testing/experience for that.

`CLEAN` does describe what's going on better, however the default value of the variable will be unset and it will be anti-idiomatic to have the _unset_ variable and the variable set to truish both perform the same action. The unset variable and setting the variable to falsey should produce the same result. `ADD` indicates that the paths should be (re)added to Spyder's runtime `sys.path`.

**Comment by Flamefire** [comment]
> `CLEAN` does describe what's going on better, however the default value of the variable will be unset and it will be anti-idiomatic to have the _unset_ variable and the variable set to truish both perform the same action. The unset variable and setting the variable to falsey should produce the same result. `ADD` indicates that the paths should be (re)added to Spyder's runtime `sys.path`.

True, I was just confused of what "ADD" would mean from the perspective of someone using `$PYTHONPATH` for that already so it sounds like it would add them twice.  
Maybe `SPY_USE_PYTHONPATH` or `SPY_ALLOW_PYTHONPATH`? Or as in `PYTHONNOUSERSITE/PYTHONDONTWRITEBYTECODE` it could be something like `SPY_NOCLEAN_PYTHONPATH`

**Comment by ccordoba12** [comment]
> So, in conclusion, I agree that @Flamefire's suggestion to use an environment variable to switch PYTHONPATH treatment in Spyder may be the best solution for now. I propose SPY_ADD_PYTHONPATH. Spyder will not add paths from PYTHONPATH to sys.path upon startup if SPY_ADD_PYTHONPATH is unset or evaluates to False; Spyder will add paths from PYTHONPATH to sys.path upon startup if SPY_ADD_PYTHONPATH evaluates to True. What do @spyder-ide/core-developers think?

I agree that that's the best we could do to allow this use case.

**Comment by mrclary** [comment]
> `` it could be something like `SPY_NOCLEAN_PYTHONPATH`

I like this. 👍🏼

**Comment by Flamefire** [comment]
Yes I can test that as time allows. So there might be a bit of delay but it isn't so urgent that it should be fine.

**Comment by dalthviz** [comment]
Note: Could this be related/should supersede https://github.com/spyder-ide/spyder/issues/21633 ?

**Comment by mrclary** [comment]
> Note: Could this be related/should supersede #21633 ?

No; I think it is related to Windows subsystem for Linux. I left a comment.
However, the proposed solution for #22681 (current issue), may provide a workaround for #21633. But I think #21633 requires a dedicated fix.

**Comment by mrclary** [comment]
@Flamefire @smax1,
I'll summarize what I think are the primary issues:
* Spyder removes `PYTHONPATH` paths from its runtime environment `sys.path`, causing Spyder to fail to launch due to its inability to find necessary packages in non-standard locations (specified in `PYTHONPATH`). Spyder-kernels may be affected in the same way.
* To pass `PYTHONPATH` on to user processes (IPython Console and others), Spyder only obtains `PYTHONPATH` from shell startup files (interactive + login) and not from the environment variables in its current launch environment, precluding possible changes from e.g. environment modules.

**Comment by Flamefire** [comment]
@mrclary Correct. And just to be clarify on the 2nd point: When started from a console on Linux PYTHONPATH from the environment [is honored](https://github.com/spyder-ide/spyder/blob/5f4c5cd326ff3b2bbf7b96b95d9ad787aa2c0bb5/spyder/utils/environ.py#L114) except paths [matching a certain pattern](https://github.com/spyder-ide/spyder/blob/5f4c5cd326ff3b2bbf7b96b95d9ad787aa2c0bb5/spyder/plugins/pythonpath/utils.py#L20-L27)

And for both we want an option to disable any filtering

**Comment by mrclary** [comment]
@Flamefire, is it possible to have a dedicated python environment for Spyder to run from?
One issue with having library files located in non-standard locations, and presumably separately maintained, is that there could arise package conflicts that affect Spyder's operation.

**Comment by mrclary** [comment]
I propose the following solution.
* Spyder will _move_ `PYTHONPATH` paths to the end of `sys.path` in its runtime environment at startup. This will allow Spyder to launch if non-standard library locations are listed in `PYTHONPATH`. Currently Spyder _removes_ these paths, however moving them to the end should provide the same protection.
* Spyder will not remove `PYTHONPATH` from the subprocess environment used to start Spyder-kenels.
* Spyder-kernels will _move_ `PYTHONPATH` paths to the end of `sys.path` at the beginning of startup. Currently, Spyder-kernels is protected when Spyder removes `PYTHONPATH` from the subprocess environment used to launch Spyder-kernels; Spyder-kernels will now be responsible for its own protection.
* Spyder-kernels will remove `PYTHONPATH` paths from `sys.path` immediately after startup. This allows proper integration with the PYTHONPATH Manager plugin. Unlike Spyder, Spyder-kernels' runtime environment is the same as the user's Python environment (where the user runs his own code) so paths from the PYTHONPATH Manager will be injected into the `sys.path` in this environment. In order to work as expected, after Spyder-kernels starts, these paths will be removed (from the end of `sys.path`) and reapplied via the `sig_pythonpath_changed` signal from the PYTHONPATH Manager plugin.
* `get_user_environment_variables` will continue to use `os.environ` from Spyder's runtime environment _if_ Spyder was launched from the commandline, otherwise it will attempt to get them from an interactive login shell subprocess. Note this is important for Spyder launched from a shortcut: macOS will not see _any_ shell startup variables, while Linux will only see login startup variables (i.e. ~/.bash_profile) but not interactive variables (i.e. ~/.bashrc).
* Add a preference "Allow unsafe PYTHONPATH paths" to Preferences -> Application -> Advanced settings (default unchecked). This will allow bypassing the path filtering performed by PYTHONPATH Manager. The `PYTHONPATH` path filtering resides entirely within the PYTHONPATH Manager, is not used anywhere else, and is unrelated to Spyder startup; therefore I think it is more appropriate to use a preference setting rather than an environment variable.

Since #21769 includes a lot of refactoring of the PYTHONPATH Manager, I propose making these changes either _after_ that PR or as part of that PR.

**Comment by mrclary** [comment]
@ccordoba12, I think we need to reconsider how Spyder obtains the environment variables. After our discussion and updating the proposed solution above, I realized that we have an issue with remote kernels.

Environment variables, and by extension PYTHONPATH paths, are currently obtained from Spyder's runtime `os.environ` or an interactive login subprocess, both of which reflect _local_ system variables. This works fine if the IPython console kernel environment is also local, but in the case of remote kernels, this is a problem. The reported environment variables from Tools -> Current user environment variables... will report environment variables from the local system, and will be different from "Show environment variables" in the IPython console hamburger menu (obtained from Spyder-kernel's `os.environ`).

Ideally, we should obtain "Current user environment variables..." _and_ "Show environment variables" from the same source: Spyder-kernels' runtime environment. This makes sense because the only _relevant_ environment variables (including `PYTHONPATH`) are those in the Spyder-kernels runtime environment. If the environment is local, then it will be effectively the same as what we currently do, but for remote kernels it will reflect the correct variables in the code execution environment.

I would propose doing the following:
* Move `spyder/utils/environ.py` to `spyder-kernels` and have Spyder-kernels handle environment retrieval operations. Thus `os.environ` or interactive login subprocesses will be derived from the kernel environment, whether local or remote.
* Remove Tools -> Current user environment variables... This does not make much sense in the context of multiple kernels that are possibly mixed local and remote. Environment variables are only relevant to the code execution environment and therefore should only be accessible from IPython Console related menus.
* PYTHONPATH Manager also obtains `PYTHONPATH` from Spyder-kernels. This may allow users to use the PYTHONPATH Manager even for remote kernels. While it may not be possible to add remote paths via navigation, the user may still be able to type in remote paths to use. Path validation could also occur by Spyder-kernels rather than Spyder prior to updating `sys.path`.

What do you think?

**Comment by Flamefire** [comment]
> @Flamefire, is it possible to have a dedicated python environment for Spyder to run from? One issue with having library files located in non-standard locations, and presumably separately maintained, is that there could arise package conflicts that affect Spyder's operation.

We more or less do have that which is why we set `PYTHONPATH` the way we do: We create an python environment with packages suitable for a given Spyder version by setting up `LD_LIBRARY_PATH`, `PYTHONPATH` and others to locations of suitable dependencies.

As all of them comes from recipes (think something like Docker or Singularity files) all software is (mostly) commonly maintained.

> Add a preference "Allow unsafe PYTHONPATH paths" to Preferences -> Application -> Advanced settings (default unchecked). This will allow bypassing the path filtering performed by PYTHONPATH Manager.

What exactly is that filtering? And what are "unsafe PYTHONPATH paths"?

**Comment by mrclary** [comment]
> What exactly is that filtering? 

I'm referring to the same thing you commented on [here](https://github.com/spyder-ide/spyder/issues/21633#issuecomment-2434586281).

> And what are "unsafe PYTHONPATH paths"?

Anything that is caught by the filtering is considered (potentially) unsafe 😄.

**Comment by mrclary** [comment]
@Flamefire @ccordoba12, I know that this issue is a year old now and we haven't addressed it yet; my apologies. In reviewing my [previous proposal](https://github.com/spyder-ide/spyder/issues/22681#issuecomment-2442487803), I think I have an improved proposal. Rather than introducing an environment variable that merely toggles Spyder's _handling_ of `PYTHONPATH`, why not introduce an environment variable that will be used _like_ `PYTHONPATH` but for Spyder's runtime environment?

* Keep Spyder's current treatment of `PYTHONPATH` by removing it from the runtime `sys.path`
* Introduce the environment variable `SPY_RUNTIME_PYTHONPATH` which, if present, will be appended to Spyder's runtime `sys.path`.
* Spyder kernels will manage its own runtime security, as previously proposed, appending `PYTHONPATH` to its `sys.path`.
* The PYTHONPATH Manager will get a preference to allow "unsafe" paths, as previously proposed.

My motivation here is to separate the intended tasks we are expecting `PYTHONPATH` to do, thereby reducing confusion and potential issues. In this way, we can allow `PYTHONPATH` to retain and honor its conventional purpose to provide additional paths for user's code execution environment, while still accommodating the need for specifying additional paths for Spyder's runtime environment.

For setups using technologies like lmod, this will also help make clear exactly which paths are used for which purpose, rather than mixing them into one environment variable. I.e. for paths required to run Spyder, use `SPY_RUNTIME_PYTHONPATH`; for paths required by the user to run code, use (as expected) `PYTHONPATH`. Either one or the other or both can then be used without interference.

@spyder-ide/core-developers, what do you think?

**Comment by ccordoba12** [comment]
> * Keep Spyder's current treatment of PYTHONPATH by removing it from the runtime sys.path
> * Introduce the environment variable SPY_RUNTIME_PYTHONPATH which, if present, will be appended to Spyder's runtime sys.path

The problem I see with this proposal is that it wouldn't be easily discoverable by users with special `PYTHONPATH` needs that must set `SPY_RUNTIME_PYTHONPATH` instead (if I understood things correctly, that's what you're proposing, right?)

**Comment by mrclary** [comment]
> The problem I see with this proposal is that it wouldn't be easily discoverable by users with special `PYTHONPATH` needs that must set `SPY_RUNTIME_PYTHONPATH` instead (if I understood things correctly, that's what you're proposing, right?)

That is correct. I think that anyone with special `PYTHONPATH` needs should be able to discover `SPY_RUNTIME_PYTHONPATH` through documentation we provide either online or through `spyder -h` cli. The benefit, as I see it, would be to avoid any possible unintended/unforeseen consequences to dual-purposing `PYTHONPATH`.

**Comment by ccordoba12** [comment]
Ok, but if we go with your proposed solution:

> Introduce the environment variable SPY_RUNTIME_PYTHONPATH which, if present, will be appended to Spyder's runtime sys.path.

then we still wouldn't cover what @Flamefire needs, i.e. that Spyder doesn't clear `PYTHONPATH` at startup since Spyder and/or its dependencies are installed in locations that are added to `PYTHONPATH`:

> We install Spyder alongside other globally installed Python packages that are in non-standard locations. To make Python find them we either add the locations to $PYTHONPATH or use a custom sitecustomize.py whose location must be in $PYTHONPATH

**Comment by mrclary** [comment]
> then we still wouldn't cover what [@Flamefire](https://github.com/Flamefire) needs, i.e. that Spyder doesn't clear `PYTHONPATH` at startup since Spyder and/or its dependencies are installed in locations that are added to `PYTHONPATH`

@Flamefire would need to use `SPY_RUNTIME_PYTHONPATH` for those paths that Spyder requires to run, but use `PYTHONPATH` for those paths that are needed for the kernel environment, i.e. for user code execution. @Flamefire is using LMod to manage environment variables, so I assume that he has a specific module for Spyder as well as other modules for other Python packages. The Spyder-specific module would need to set `SPY_RUNTIME_PYTHONPATH`, but would not set `PYTHONPATH`, while any other Python-specific modules intended for user code execution would set `PYTHONPATH`. These two variables may or may not have similar paths, but they don't _have_ to be the same, and that would be the whole point.

**Comment by mrclary** [comment]
@Flamefire @ccordoba12,
Are there any further comments regarding my last proposed solution?
Should this issue remain open?
